src/lib/optimizer/strategy/join_to_predicate_rewrite_rule.cpp

```diff
@@ -0,0 +1,200 @@
+#include "join_to_predicate_rewrite_rule.hpp"
+
+#include "expression/abstract_expression.hpp"
+#include "expression/binary_predicate_expression.hpp"
+#include "expression/expression_functional.hpp"
+#include "expression/expression_utils.hpp"
+#include "expression/lqp_column_expression.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "magic_enum.hpp"
+
+namespace opossum {
+
+std::string JoinToPredicateRewriteRule::name() const {
+  static const auto name = std::string{"JoinToPredicateRewriteRule"};
+  return name;
+}
+
+void JoinToPredicateRewriteRule::_apply_to_plan_without_subqueries(
+    const std::shared_ptr<AbstractLQPNode>& lqp_root) const {
+  auto rewritable_nodes = std::vector<std::shared_ptr<JoinNode>>();
+  auto removable_sides = std::vector<LQPInputSide>();
+  auto valid_predicates = std::vector<std::shared_ptr<PredicateNode>>();
```

Comment:
Maybe store one vector of structs (instead of three separate vectors) to be more explicit about the fact that the elements at the same index belong together?

Reply:
Wouldn't a `vector<std::tuple<JoinNode, InputSide, PNode>>` be even better?

Easier readable (imo, just a single index access) and no further data structures required.